### PR TITLE
Draft+ for SDXL  [draft]

### PR DESCRIPTION
**Pull request by rohitrango**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

# Changelog 
- Please update the [CHANGELOG.md](/CHANGELOG.md) under next version with high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation? Make sure to also update the [NeMo Framework User Guide](https://docs.nvidia.com/nemo-framework/user-guide/latest/index.html) which contains the tutorials

# Checklist when contributing a new algorithm
- [ ] Does the trainer resume and restore model state all states?
- [ ] Does the trainer support all parallelism techniques(PP, TP, DP)?
- [ ] Does the trainer support `max_steps=-1` and `validation`?
- [ ] Does the trainer only call APIs defined in [alignable_interface.py](/nemo_aligner/models/alignable_interface.py)?
- [ ] Does the trainer have proper logging?

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo_aligner/algorithms/supervised.py [hunk 1]
```diff
@@ -181,6 +183,15 @@ def fit(self):
 
         self.run_timer.start_time()
 
+        if self.run_init_validation:
+            logging.info("Running validation in the very beginning.")
+            # run an initial validation
```
Comment (terrykong) [inline nemo_aligner/algorithms/supervised.py:188]: Comment seems redundant with the log line right above

File: nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+        """things to call to prepare for validation
+        """
+        prepare_for_validation_step(self)
+        gbs = int(self.cfg.global_batch_size)
+        mbs = int(self.cfg.micro_batch_size)
+        dp_size = int(parallel_state.get_data_parallel_world_size())
+        configure_batch_sizes(mbs=mbs, gbs=gbs, dp=dp_size)
+
+    def finish_validation_step(self):
+        """things to call to prepare for validation
```
Comment (terrykong) [inline nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py:215]: This comment doesn't reflect method name

File: nemo_aligner/utils/train_script_utils.py [hunk 3]
```diff
@@ -101,6 +102,7 @@ def init_using_ptl(ptl_trainer, ptl_model, train_dataloader, train_ds):
     disable_data_callbacks(ptl_model, train_dataloader, train_ds)
 
     call._call_setup_hook(ptl_trainer)
+    call._call_configure_model(ptl_trainer)
```
Comment (terrykong) [inline nemo_aligner/utils/train_script_utils.py:105]: For my education, does this fix anything? It looks like in [PTL](https://github.com/Lightning-AI/pytorch-lightning/blob/2524864b3cc8dae3552a4b5c3c819c2ce6f278af/src/lightning/pytorch/trainer/trainer.py#L950), this is indeed called after `_call_setup_hook`, but just wondering how we got away with not calling this

File: nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+            x = x_T + 0
+            num_steps = sampler.num_steps
+            x, s_in, sigmas, num_sigmas, cond, uc = sampler.prepare_sampling_loop(x, cond, uc, num_steps)
+            # last step doesnt count since there is no additional sigma
+            total_steps = num_sigmas - 1
+
+            iterator = tqdm(range(num_sigmas - 1), desc=f"{sampler.__class__.__name__} Sampler", total=total_steps)
+            for i in iterator:
+                gamma = sampler.get_gamma(sigmas, num_sigmas, i)
+                # with context(set_draft_grad_flag):
```
Comment (terrykong) [inline nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py:358]: Commented out code needed?

File: nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+            # TODO: Currently the main parameter data type is kept in fp32 (when O2=False). This needs to be
+            # extended to support lower precision main parameters.
+            frozen_submodule_names, frozen_submodules = find_frozen_submodules(self.model)
+            for submodule in frozen_submodule_names:
+                logging.debug(f"Ignoring state {submodule} in FSDP.")
+            self.trainer.strategy.kwargs["ignored_states"] = frozen_submodules
+            # FSDP requires uniform status of require_grads
+            # Diffusion models like SD has frozen parts and needs to be added to 'ignored_states' from sharding for FSDP to work
+            self.model = self.trainer.strategy._setup_model(self.model)
+            # self.model.diffusion_model.first_stage = FullyShardedDataParallel
```
Comment (terrykong) [inline nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py:184]: Commented out code needed?

**Review by terrykong — CHANGES_REQUESTED** [review]
Thanks @rohitrango for the contribution!

I've left some comments. Let me know if there are any questions

File: examples/mm/stable_diffusion/launch_draft_oci_multinode_xl.sh
[the remark this answers is not in the document]
Comment (rohitrango) [reply examples/mm/stable_diffusion/launch_draft_oci_multinode_xl.sh:1]: Yes! Removing this from here.

File: .gitignore [hunk 6]
```diff
@@ -1,4 +1,5 @@
 # log and data files
+hub
```
[the remark this answers is not in the document]
Comment (rohitrango) [reply .gitignore:2]: Makes sense.

File: examples/mm/stable_diffusion/conf/draftp_sdxl.yaml
[the remark this answers is not in the document]
Comment (rohitrango) [reply examples/mm/stable_diffusion/conf/draftp_sdxl.yaml:1]: Done. I guess I can just delete the extra configs.

File: examples/mm/stable_diffusion/train_sdxl_draftp.py [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+def resolve_and_create_trainer(cfg, pop_trainer_key):
+    """resolve the cfg, remove the key before constructing the PTL trainer
+        and then restore it after
+    """
+    OmegaConf.resolve(cfg)
+    with temp_pop_from_config(cfg.trainer, pop_trainer_key):
+        return MegatronStableDiffusionTrainerBuilder(cfg).create_trainer()
+
+
+def wandb_login():
```
[the remark this answers is not in the document]
Comment (rohitrango) [reply examples/mm/stable_diffusion/train_sdxl_draftp.py:151]: True! this was a hack because I had some other env variable set for wandb. deleting this

File: nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py [hunk 8]
```diff
@@ -0,0 +1,500 @@
...
+
+    return kl
+
+
+class MegatronSDXLDRaFTPModel(MegatronDiffusionEngine, SupervisedInterface):
+    def __init__(self, cfg, trainer):
+
+        super().__init__(cfg, trainer=trainer)
+        # init_cfg = deepcopy(cfg)
+        # init_cfg.peft.peft_scheme = "none"
```
[the remark this answers is not in the document]
Comment (rohitrango) [reply nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py:79]: no... this was while I was debugging. removing this

File: nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py (hunk 2 above)
Comment (rohitrango) [reply nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py:215]: I actually copied this from `megatron_sd_draftp_model`. I'll add the correct comment in both files.

File: nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py [hunk 9]
```diff
@@ -0,0 +1,500 @@
...
+                    batch_size,
+                    self.in_channels,
+                    self.height // self.downsampling_factor,
+                    self.width // self.downsampling_factor,
+                ],
+                generator=None,
+            ).to(torch.cuda.current_device())
+
+        image_draft_p, reward_draft_p, vae_decoder_output_draft_p = self.generate_log_images(
+            latents + 0, prompts, self.model
```
[the remark this answers is not in the document]
Comment (rohitrango) [reply nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py:293]: it just creates a new copy of the tensor in case its modified by some in-place op inside the function. 

I guess I could also use `.clone()` function, that'd be cleaner.

File: nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py [hunk 10]
```diff
@@ -0,0 +1,500 @@
...
+                            return_noise=True,
+                        )
+                        x = x_next_draft
+                else:
+                    # store computation graph of draft eps
+                    x_next_draft, eps_draft = sampler.sampler_step(
+                        s_in * sigmas[i], s_in * sigmas[i + 1], denoiser_draft, x, cond, uc, gamma, return_noise=True
+                    )
+                    list_eps_draft.append(eps_draft)
+                    ##### TODO: uncomment this for base model
```
[the remark this answers is not in the document]
Comment (rohitrango) [reply nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py:379]: already commented out lol, removing todo

File: nemo_aligner/utils/train_script_utils.py (hunk 3 above)
Comment (rohitrango) [reply nemo_aligner/utils/train_script_utils.py:105]: Yes, this is very tricky. The PTL calls are 
```
- strategy.connect(model)
- strategy.setup_environment()
- call._call_setup_hook(self)
- call._call_configure_model(self)
- self.strategy.setup(self)
- _run_model [this is the actual training loop]
```

this function actually calls `configure_sharded_model` of the Megatron model which is a no-op if FSDP is off (check `megatron_base_model.py`). If FSDP is on, then this function calls `FullyShardedDataParallel(model, ...)` instead of the lightning trainer calling it. 

Without this function, FSDP does not work. I guess it worked so far because no one tested FSDP functionality in aligner? @JRD971000 told me none of the other aligner algos use fsdp. Is that true?

File: nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py [hunk 11]
```diff
@@ -0,0 +1,479 @@
...
+
+        if self.use_fsdp:
+            """Top-evel FSDP model sharding"""
+            # Shard the top-level model hierarchically. We shard the strategy-unwrapped model not
+            # to lose the structure of non-FSDP wrapped parameters (e.g, embedding)
+            # TODO: Currently the main parameter data type is kept in fp32 (when O2=False). This needs to be
+            # extended to support lower precision main parameters.
+            frozen_submodule_names, frozen_submodules = find_frozen_submodules(self.model)
+            for submodule in frozen_submodule_names:
+                logging.debug(f"Ignoring state {submodule} in FSDP.")
```
Comment (terrykong) [inline nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py:172]: This sounds more like a `logging.warning`

File: nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py
Comment (terrykong) [inline nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py:1]: This file seems pretty close to  the existing sd one, with the exception of sampling looking like it comes from a different API. Why not have a single loop if the overlap is so high?

**Review by terrykong — CHANGES_REQUESTED** [review]

File: examples/mm/stable_diffusion/train_sdxl_draftp.py [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+def main(cfg) -> None:
+
+    logging.info("\n\n************** Experiment configuration ***********")
+    logging.info(f"\n{OmegaConf.to_yaml(cfg)}")
+
+    # set cuda device for each process
+    local_rank = int(os.environ.get("LOCAL_RANK", 0))
+    torch.cuda.set_device(local_rank)
+
+    # TODO: has to be set true for PyTorch 1.12 and later.
```
Comment (terrykong) [inline examples/mm/stable_diffusion/train_sdxl_draftp.py:160]: Are we testing on pytorch that old? If not can we remove comment?

But if we are, then we should should guard this (w/ packaging.verion.Version):
```suggestion
    if Version(torch.__version__) > Version('1.12'):
        torch.backends.cuda.matmul.allow_tf32 = True
```

File: examples/mm/stable_diffusion/train_sdxl_draftp.py [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+    optimizer, scheduler = extract_optimizer_scheduler_from_ptl_model(ptl_model)
+
+    ckpt_callback = add_custom_checkpoint_callback(trainer, ptl_model)
+
+    logger.log_hyperparams(OmegaConf.to_container(cfg))
+
+    if local_rank == 0:
+        print(ptl_model)
+        # import time
+        # time.sleep(10)
```
Comment (terrykong) [inline examples/mm/stable_diffusion/train_sdxl_draftp.py:240]: Can we clean up?
```suggestion
```

File: nemo_aligner/utils/train_script_utils.py (hunk 3 above)
Comment (terrykong) [reply nemo_aligner/utils/train_script_utils.py:105]: Checked with Gerald and this change SG

File: nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py
Comment (terrykong) [reply nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py:1]: Resolved offline. The classes for SD and SDXL are different even in Nemo, so the unification of the conditioner API should happen in nemo for us to take advantage of a simpler API. I created an issue to track

File: examples/mm/stable_diffusion/train_sdxl_draftp.py (hunk 12 above)
Comment (rohitrango) [reply examples/mm/stable_diffusion/train_sdxl_draftp.py:160]: not really. I've tested everything on torch2.3.0. I'ts a comment I picked up from `train_sd_draftp.py` -- adding guard

File: examples/mm/stable_diffusion/train_sdxl_draftp.py [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+
+def resolve_and_create_trainer(cfg, pop_trainer_key):
+    """resolve the cfg, remove the key before constructing the PTL trainer
+        and then restore it after
+    """
+    OmegaConf.resolve(cfg)
+    with temp_pop_from_config(cfg.trainer, pop_trainer_key):
+        return MegatronStableDiffusionTrainerBuilder(cfg).create_trainer()
+
+
```
Comment (terrykong) [inline examples/mm/stable_diffusion/train_sdxl_draftp.py:150]: @rohitrango Do you mind leaving a comment like
```suggestion

# TODO(@rohitrango): Merge script with SD train script after models/APIs in Nemo are merged
```
This way we won't forget that to deduplicate this code once Nemo APIs are merged

File: examples/mm/stable_diffusion/conf/draftp_sdxl.yaml [hunk 15]
```diff
@@ -0,0 +1,281 @@
+name: sdxl_draftp_train
```
Comment (terrykong) [inline examples/mm/stable_diffusion/conf/draftp_sdxl.yaml:1]: @rohitrango Do you mind leaving a comment like
```suggestion

# TODO(@rohitrango): Remove after SD/SDXL APIs merged in NeMo so only one config is needed in Aligner
```
This way we won't forget that to deduplicate this code once Nemo APIs are merged

File: nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py [hunk 16]
```diff
@@ -0,0 +1,478 @@
...
+
+def calculate_gaussian_kl_penalty_shared_var(curr_eps, init_eps):
+    diff = curr_eps - init_eps
+    kl = torch.sum(diff ** 2, dim=(1, 2, 3, 4), keepdim=True).flatten()
+    dimensionality = torch.numel(curr_eps[0])
+    kl /= dimensionality
+    return kl
+
+
+class MegatronSDXLDRaFTPModel(MegatronDiffusionEngine, SupervisedInterface):
```
Comment (terrykong) [inline nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py:73]: @rohitrango Do you mind leaving a comment like
```suggestion

# TODO(@rohitrango): Remove after SD/SDXL APIs merged in NeMo so only one config is needed in Aligner
```
This way we won't forget that to deduplicate this code once Nemo APIs are merged

File: nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (terrykong) [reply nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py]: Yea, I like your proposal of using .clone()

**Review by yaoyu-33 — APPROVED** [review]

**Comment by terrykong** [comment]
Requires https://github.com/NVIDIA/NeMo/pull/9543

**Review by terrykong — APPROVED** [review]
Thanks @rohitrango for adding this new feature!
